### PR TITLE
:warning: Bump to Kubernetes 1.17 dependencies

### PR DESCRIPTION
**Pull request by joelanford**
This PR bumps controller-runtime to use Kubernetes 1.17 libraries, which will enable controller/operator developers to take advantage of the latest 1.17 client-go and API updates in their projects.

Due to an upstream change in `k8s.io/api`, there is one breaking change in controller-runtime's `pkg/leaderelection/fake` package, which is used only in `pkg/manager/manager_test.go`, at least internally. Should we deprecate this fake ResourceLock implementation from the public API and plan to move this into an `internal` tree?

Here's the notes and output from `go-apidiff`:

- There's the one aforementioned breaking change in `sigs.k8s.io/controller-runtime`
- As is expected with a version bump to the next major Kubernetes version, there are a raft of breaking API changes to libraries that controller-runtime imports and uses. For users of controller-runtime that also use these libraries, there may need to be code refactorings to address these breakages.

```
$ go-apidiff master k8s-1.17 --compare-imports

sigs.k8s.io/controller-runtime/pkg/leaderelection/fake
  Incompatible changes:
  - (*ResourceLock).Get: changed from func() (*k8s.io/client-go/tools/leaderelection/resourcelock.LeaderElectionRecord, error) to func() (*k8s.io/client-go/tools/leaderelection/resourcelock.LeaderElectionRecord, []byte, error)

k8s.io/client-go/kubernetes
  Incompatible changes:
  - Interface.DiscoveryV1beta1: added
  - Interface.FlowcontrolV1alpha1: added

github.com/onsi/ginkgo/config
  Incompatible changes:
  - VERSION: value changed from "1.8.0" to "1.10.1"

k8s.io/apimachinery/pkg/apis/meta/v1
  Incompatible changes:
  - Convert_Slice_string_To_v1_DeletionPropagation: removed

k8s.io/client-go/rest
  Incompatible changes:
  - (*Request).WatchWithSpecificDecoders: removed
  - NewRESTClient: changed from func(*net/url.URL, string, ContentConfig, float32, int, k8s.io/client-go/util/flowcontrol.RateLimiter, *net/http.Client) (*RESTClient, error) to func(*net/url.URL, string, ClientContentConfig, k8s.io/client-go/util/flowcontrol.RateLimiter, *net/http.Client) (*RESTClient, error)
  - NewRequest: changed from func(HTTPClient, string, *net/url.URL, string, ContentConfig, Serializers, BackoffManager, k8s.io/client-go/util/flowcontrol.RateLimiter, time.Duration) *Request to func(*RESTClient) *Request
  - RESTClient.Throttle: removed
  - Serializers: removed

github.com/prometheus/client_golang/prometheus
  Incompatible changes:
  - DefObjectives: removed
  - Handler: removed
  - InstrumentHandler: removed
  - InstrumentHandlerFunc: removed
  - InstrumentHandlerFuncWithOpts: removed
  - InstrumentHandlerWithOpts: removed
  - UninstrumentedHandler: removed

k8s.io/client-go/tools/leaderelection/resourcelock
  Incompatible changes:
  - (*ConfigMapLock).Get: changed from func() (*LeaderElectionRecord, error) to func() (*LeaderElectionRecord, []byte, error)
  - (*EndpointsLock).Get: changed from func() (*LeaderElectionRecord, error) to func() (*LeaderElectionRecord, []byte, error)
  - (*LeaseLock).Get: changed from func() (*LeaderElectionRecord, error) to func() (*LeaderElectionRecord, []byte, error)
  - Interface.Get: changed from func() (*LeaderElectionRecord, error) to func() (*LeaderElectionRecord, []byte, error)

k8s.io/apimachinery/pkg/labels
  Incompatible changes:
  - Selector.RequiresExactMatch: added

k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
  Incompatible changes:
  - JSONFallbackEncoder: removed

github.com/onsi/ginkgo
  Incompatible changes:
  - GINKGO_VERSION: value changed from "1.8.0" to "1.10.1"

k8s.io/api/core/v1
  Incompatible changes:
  - NodeSpec.DoNotUse_ExternalID: removed

k8s.io/client-go/tools/record
  Incompatible changes:
  - EventBroadcaster.Shutdown: added

k8s.io/apimachinery/pkg/runtime
  Incompatible changes:
  - Encoder.Identifier: added
  - GroupVersioner.Identifier: added
  - GroupVersioners: removed
  - VersionedObjects: removed
```

### CODE REVIEW

File: pkg/cache/internal/deleg_map.go [hunk 1]
```diff
@@ -68,6 +68,10 @@ func (m *InformersMap) WaitForCacheSync(stop <-chan struct{}) bool {
 	syncedFuncs := append([]cache.InformerSynced(nil), m.structured.HasSyncedFuncs()...)
 	syncedFuncs = append(syncedFuncs, m.unstructured.HasSyncedFuncs()...)
 
+	// TODO(jlanford): investigate why this sleep is necessary
+	//   to keep the tests passing. This is related to
+	//   kubernetes/client-go@6b59ea5
+	time.Sleep(time.Millisecond * 100)
```
Comment (joelanford) [inline pkg/cache/internal/deleg_map.go]: This code can be removed when #755 is merged.

Comment (DirectXMan12) [reply pkg/cache/internal/deleg_map.go]: We def should not merge with this in place

**Comment by joelanford** [comment]
/hold

Wait until #755 is merged and this is rebased.

File: pkg/cache/internal/deleg_map.go (hunk 1 above)
Comment (joelanford) [reply pkg/cache/internal/deleg_map.go]: @DirectXMan12 this has been removed now that #755 is merged.

**Comment by gerred** [comment]
/assign @gerred 
/assign @vincepri 

We're going to do this in a pipeline, get 1.16.4 out, then move on to 1.17 in short order -- and ideally automate this to reduce toil @joelanford.

**Comment by gerred** [comment]
In theory though, this LGTM, but going to start work on getting this into a full merge to release train.

**Comment by vincepri** [comment]
@joelanford Can you please squash commits and rebase?

**Comment by gerred** [comment]
I'll approve as soon as that's done.

**Comment by ialidzhikov** [comment]
> @joelanford Can you please squash commits and rebase?

ping @joelanford

**Comment by vincepri** [comment]
@gerred @ialidzhikov @joelanford Opened https://github.com/kubernetes-sigs/controller-runtime/pull/795 with squashed commits and rebased on master.

Closing this one.

/close

**Comment by joelanford** [comment]
Sorry all! GitHub notifications fail. In the future, feel free to ping me on the Kubernetes slack if I'm not responding here.

**Pull request closed without merging**